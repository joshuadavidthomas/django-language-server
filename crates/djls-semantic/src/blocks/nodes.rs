use djls_source::Span;
use serde::Serialize;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize)]
pub struct BlockId(u32);

impl BlockId {
    pub fn new(id: u32) -> Self {
        Self(id)
    }

    pub fn id(self) -> u32 {
        self.0
    }

    pub fn index(self) -> usize {
        self.0 as usize
    }
}

#[derive(Clone, Debug, Default, Serialize)]
pub struct Blocks(Vec<Region>);

impl Blocks {
    pub fn get(&self, id: usize) -> &Region {
        &self.0[id]
    }
}

impl IntoIterator for Blocks {
    type Item = Region;
    type IntoIter = std::vec::IntoIter<Region>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}

impl<'a> IntoIterator for &'a Blocks {
    type Item = &'a Region;
    type IntoIter = std::slice::Iter<'a, Region>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.iter()
    }
}

impl<'a> IntoIterator for &'a mut Blocks {
    type Item = &'a mut Region;
    type IntoIter = std::slice::IterMut<'a, Region>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.iter_mut()
    }
}

impl Blocks {
    pub fn alloc(&mut self, span: Span, parent: Option<BlockId>) -> BlockId {
        let id = BlockId(self.0.len() as u32);
        self.0.push(Region::new(span, parent));
        id
    }

    pub fn extend_block(&mut self, id: BlockId, span: Span) {
        self.block_mut(id).extend_span(span);
    }

    pub fn push_node(&mut self, target: BlockId, node: BlockNode) {
        let span = node.span();
        self.extend_block(target, span);
        self.block_mut(target).nodes.push(node);
    }

    pub fn block_mut(&mut self, id: BlockId) -> &mut Region {
        let idx = id.index();
        &mut self.0[idx]
    }
}

#[derive(Clone, Debug, Serialize)]
pub struct Region {
    span: Span,
    nodes: Vec<BlockNode>,
    parent: Option<BlockId>,
}

impl Region {
    fn new(span: Span, parent: Option<BlockId>) -> Self {
        Self {
            span,
            nodes: Vec::new(),
            parent,
        }
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn set_span(&mut self, span: Span) {
        self.span = span;
    }

    pub fn nodes(&self) -> &Vec<BlockNode> {
        &self.nodes
    }

    fn extend_span(&mut self, span: Span) {
        let opening = self.span.start().saturating_sub(span.start());
        let closing = span.end().saturating_sub(self.span.end());
        self.span = self.span.expand(opening, closing);
    }
}

#[derive(Clone, Debug, Serialize)]
pub enum BranchKind {
    Opener,
    Segment,
}

#[derive(Clone, Debug, Serialize)]
pub enum BlockNode {
    Leaf {
        label: String,
        span: Span,
    },
    Branch {
        tag: String,
        marker_span: Span,
        body: BlockId,
        kind: BranchKind,
    },
    Error {
        message: String,
        span: Span,
    },
}

impl BlockNode {
    fn span(&self) -> Span {
        match self {
            BlockNode::Leaf { span, .. } | BlockNode::Error { span, .. } => *span,
            BlockNode::Branch { marker_span, .. } => *marker_span,
        }
    }
}
