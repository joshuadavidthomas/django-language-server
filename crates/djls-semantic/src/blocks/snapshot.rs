use djls_source::Span;
use serde::Serialize;
use std::collections::HashSet;

use super::nodes::{BlockId, BlockNode, BranchKind};
use super::tree::BlockTree;

#[derive(Serialize)]
pub struct BlockTreeSnapshot {
    roots: Vec<u32>,
    root_ids: Vec<u32>,
    blocks: Vec<BlockSnapshot>,
}

impl From<&BlockTree> for BlockTreeSnapshot {
    #[allow(clippy::too_many_lines)]
    fn from(tree: &BlockTree) -> Self {
        let mut container_ids: HashSet<u32> = HashSet::new();
        let mut body_ids: HashSet<u32> = HashSet::new();

        for r in tree.roots() {
            container_ids.insert(r.id());
        }
        for (i, b) in tree.blocks().into_iter().enumerate() {
            let i_u = u32::try_from(i).unwrap_or(u32::MAX);
            for n in b.nodes() {
                match n {
                    BlockNode::Leaf { .. } | BlockNode::Error { .. } => {}
                    BlockNode::Branch {
                        body,
                        kind: BranchKind::Opener,
                        ..
                    } => {
                        container_ids.insert(body.id());
                    }
                    BlockNode::Branch {
                        body,
                        kind: BranchKind::Segment,
                        ..
                    } => {
                        body_ids.insert(body.id());
                    }
                }
            }
            if container_ids.contains(&i_u) {
                body_ids.remove(&i_u);
            }
        }

        let blocks = tree
            .blocks()
            .into_iter()
            .enumerate()
            .map(|(i, b)| {
                let id_u = u32::try_from(i).unwrap_or(u32::MAX);
                let nodes: Vec<BlockNodeSnapshot> = b
                    .nodes()
                    .iter()
                    .map(|n| match n {
                        BlockNode::Leaf { label, span } => BlockNodeSnapshot::Leaf {
                            label: label.clone(),
                            span: *span,
                        },
                        BlockNode::Error { message, span } => BlockNodeSnapshot::Error {
                            message: message.clone(),
                            span: *span,
                        },
                        BlockNode::Branch {
                            tag,
                            marker_span,
                            body,
                            ..
                        } => BlockNodeSnapshot::Branch {
                            block_id: body.id(),
                            tag: tag.clone(),
                            marker_span: *marker_span,
                            content_span: *tree.blocks().get(body.index()).span(),
                        },
                    })
                    .collect();

                if container_ids.contains(&id_u) {
                    BlockSnapshot::Container {
                        container_span: *b.span(),
                        nodes,
                    }
                } else {
                    BlockSnapshot::Body {
                        content_span: *b.span(),
                        nodes,
                    }
                }
            })
            .collect();

        // Also compute root_id for every block/region
        let root_ids: Vec<u32> = tree
            .blocks()
            .into_iter()
            .enumerate()
            .map(|(i, _)| {
                let mut cur = BlockId::new(u32::try_from(i).unwrap_or(u32::MAX));
                // climb via snapshot-internal parent pointers
                loop {
                    // safety: we have no direct parent access in snapshot; infer by scanning containers
                    // If any Branch points to `cur` as body, that region's parent is its container id
                    let mut parent: Option<BlockId> = None;
                    for (j, b) in tree.blocks().into_iter().enumerate() {
                        for n in b.nodes() {
                            if let BlockNode::Branch { body, .. } = n {
                                if body.index() == cur.index() {
                                    parent =
                                        Some(BlockId::new(u32::try_from(j).unwrap_or(u32::MAX)));
                                    break;
                                }
                            }
                        }
                        if parent.is_some() {
                            break;
                        }
                    }
                    if let Some(p) = parent {
                        cur = p;
                    } else {
                        break cur.id();
                    }
                }
            })
            .collect();

        Self {
            roots: tree.roots().iter().map(|r| r.id()).collect(),
            blocks,
            root_ids,
        }
    }
}

#[derive(Serialize)]
#[serde(tag = "kind")]
pub enum BlockSnapshot {
    Container {
        container_span: Span,
        nodes: Vec<BlockNodeSnapshot>,
    },
    Body {
        content_span: Span,
        nodes: Vec<BlockNodeSnapshot>,
    },
}

#[derive(Serialize)]
#[serde(tag = "node")]
pub enum BlockNodeSnapshot {
    Branch {
        block_id: u32,
        tag: String,
        marker_span: Span,
        content_span: Span,
    },
    Leaf {
        label: String,
        span: Span,
    },
    Error {
        message: String,
        span: Span,
    },
}
