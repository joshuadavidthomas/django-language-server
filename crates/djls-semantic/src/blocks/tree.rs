use serde::Serialize;

use super::nodes::BlockId;
use super::nodes::Blocks;

#[derive(Clone, Debug, Serialize)]
pub struct BlockTree {
    roots: Vec<BlockId>,
    blocks: Blocks,
}

impl BlockTree {
    pub fn new() -> Self {
        Self {
            roots: Vec::new(),
            blocks: Blocks::default(),
        }
    }

    pub fn roots(&self) -> &Vec<BlockId> {
        &self.roots
    }

    pub fn roots_mut(self) -> Vec<BlockId> {
        self.roots
    }

    pub fn blocks(&self) -> &Blocks {
        &self.blocks
    }

    pub fn blocks_mut(self) -> Blocks {
        self.blocks
    }
}

impl Default for BlockTree {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::blocks::snapshot::BlockTreeSnapshot;
    use crate::{templatetags::django_builtin_specs, TagSpecs};
    use camino::Utf8Path;
    use djls_source::File;
    use djls_templates::parse_template;
    use djls_workspace::FileSystem;
    use djls_workspace::InMemoryFileSystem;
    use std::sync::Arc;
    use std::sync::Mutex;

    impl BlockTree {
        pub fn to_snapshot(&self) -> BlockTreeSnapshot {
            BlockTreeSnapshot::from(self)
        }
    }

    #[salsa::db]
    #[derive(Clone)]
    struct TestDatabase {
        storage: salsa::Storage<Self>,
        fs: Arc<Mutex<InMemoryFileSystem>>,
    }

    impl TestDatabase {
        fn new() -> Self {
            Self {
                storage: salsa::Storage::default(),
                fs: Arc::new(Mutex::new(InMemoryFileSystem::new())),
            }
        }

        fn add_file(&self, path: &str, content: &str) {
            self.fs
                .lock()
                .unwrap()
                .add_file(path.into(), content.to_string());
        }
    }

    #[salsa::db]
    impl salsa::Database for TestDatabase {}

    #[salsa::db]
    impl djls_source::Db for TestDatabase {
        fn read_file_source(&self, path: &Utf8Path) -> std::io::Result<String> {
            self.fs.lock().unwrap().read_to_string(path)
        }
    }

    #[salsa::db]
    impl djls_templates::Db for TestDatabase {}

    #[salsa::db]
    impl crate::db::Db for TestDatabase {
        fn tag_specs(&self) -> TagSpecs {
            django_builtin_specs()
        }
    }

    #[test]
    fn test_block_tree_building() {
        let db = TestDatabase::new();

        let source = r"
{% block header %}
    <h1>Title</h1>
{% endblock header %}

{% if user.is_authenticated %}
    <p>Welcome {{ user.name }}</p>
    {% if user.is_superuser %}
        <span>Admin</span>
    {% elif user.is_staff %}
        <span>Manager</span>
    {% else %}
        <span>Regular user</span>
    {% endif %}
{% else %}
    <p>Please log in</p>
{% endif %}

{% for item in items %}
    <li>{{ item }}</li>
{% endfor %}
";

        db.add_file("test.html", source);
        let file = File::new(&db, "test.html".into(), 0);
        let nodelist = parse_template(&db, file).expect("should parse");

        let nodelist_view = {
            #[derive(serde::Serialize)]
            struct NodeListView {
                nodes: Vec<NodeView>,
            }
            #[derive(serde::Serialize)]
            #[serde(tag = "kind")]
            enum NodeView {
                Tag {
                    name: String,
                    bits: Vec<String>,
                    span: Span,
                },
                Variable {
                    var: String,
                    filters: Vec<String>,
                    span: Span,
                },
                Comment {
                    content: String,
                    span: Span,
                },
                Text {
                    span: Span,
                },
                Error {
                    span: Span,
                    full_span: Span,
                    error: String,
                },
            }

            let nodes = nodelist
                .nodelist(&db)
                .iter()
                .map(|n| match n {
                    Node::Tag { name, bits, span } => NodeView::Tag {
                        name: name.text(&db).to_string(),
                        bits: bits.iter().map(|b| b.text(&db).to_string()).collect(),
                        span: *span,
                    },
                    Node::Variable { var, filters, span } => NodeView::Variable {
                        var: var.text(&db).to_string(),
                        filters: filters.iter().map(|f| f.text(&db).to_string()).collect(),
                        span: *span,
                    },
                    Node::Comment { content, span } => NodeView::Comment {
                        content: content.clone(),
                        span: *span,
                    },
                    Node::Text { span } => NodeView::Text { span: *span },
                    Node::Error {
                        span,
                        full_span,
                        error,
                    } => NodeView::Error {
                        span: *span,
                        full_span: *full_span,
                        error: error.to_string(),
                    },
                })
                .collect();

            NodeListView { nodes }
        };
        insta::assert_yaml_snapshot!("nodelist", nodelist_view);
        let tag_shapes = TagShapes::from(&db.tag_specs());
        let block_tree = BlockTree::build(&db, nodelist, &tag_shapes);
        insta::assert_yaml_snapshot!("blocktree", block_tree.to_snapshot());
    }
}
