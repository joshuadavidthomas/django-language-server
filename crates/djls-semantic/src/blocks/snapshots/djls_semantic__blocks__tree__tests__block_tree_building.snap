---
source: crates/djls-semantic/src/blocks/tree.rs
expression: "TreeSnapshot::capture(&db, block_tree)"
---
roots:
  - root_idx: 0
    blocks:
      - Branch:
          tag: block
          marker_span:
            - 3
            - 17
          kind: Segment
          nested_blocks:
            - Leaf:
                label: "<var>"
                span:
                  - 30
                  - 37
  - root_idx: 2
    blocks:
      - Branch:
          tag: if
          marker_span:
            - 74
            - 100
          kind: Segment
          nested_blocks:
            - Leaf:
                label: "<var>"
                span:
                  - 124
                  - 135
      - Branch:
          tag: else
          marker_span:
            - 147
            - 157
          kind: Segment
          nested_blocks: []
  - root_idx: 5
    blocks:
      - Branch:
          tag: for
          marker_span:
            - 210
            - 229
          kind: Segment
          nested_blocks:
            - Leaf:
                label: "<var>"
                span:
                  - 247
                  - 253
      - Branch:
          tag: empty
          marker_span:
            - 266
            - 277
          kind: Segment
          nested_blocks: []
  - root_idx: 8
    blocks:
      - Branch:
          tag: comment
          marker_span:
            - 334
            - 343
          kind: Segment
          nested_blocks: []
